File: docs/prefix_sum/2302_hard/subarray_scores.py

def count_subarray_scores(nums: list[int], k: int) -> int:  # LeetCode Q.2302.
    total_subarrays = 0
    subarray_sum = 0

    start_idx = 0
    for end_idx, num in enumerate(nums):
        subarray_sum += nums[end_idx]
        score = subarray_sum * (end_idx + 1 - start_idx)

        while score >= k and start_idx <= end_idx:
            subarray_sum -= nums[start_idx]
            start_idx += 1
            score = subarray_sum * (end_idx + 1 - start_idx)

        total_subarrays += end_idx + 1 - start_idx

    return total_subarrays

File: docs/prefix_sum/2302_hard/subarray_scores.cpp
#include <vector>
using namespace std;

long long countSubarrayScores(vector<int>& nums, long long k) // LeetCode Q.2302.
{
    long long lessThanKSubarraysCount = 0; // Long long prevents overflow.

    long long subarraySum = 0; // Long long prevents overflow.

    int leftIdx = 0;
    for (int rightIdx = 0; rightIdx < nums.size(); rightIdx++) {
        subarraySum += nums[rightIdx];

        while (subarraySum * (rightIdx + 1 - leftIdx) >= k) {
            subarraySum -= nums[leftIdx];
            leftIdx++;
        }

        lessThanKSubarraysCount += rightIdx + 1 - leftIdx;
    }

    return lessThanKSubarraysCount;
}
